File: python/2471. Minimum Number of Operations to Sort a Binary Tree by Level.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minimumOperations(self, root: Optional[TreeNode]) -> int:
        q = deque([root])
        answer = 0
        while q:
            size = len(q)
            level = []
            while size > 0:
                current = q.popleft()
                level.append(current.val)
                if current.left:
                    q.append(current.left)
                if current.right:
                    q.append(current.right)
                size -= 1
            answer += self.get_swaps(level)
        return answer
    def get_swaps(self, original: list) -> int:
        swaps = 0
        target = sorted(original)
        pos = {val: idx for idx, val in enumerate(original)}
        for i in range(len(original)):
            if original[i] != target[i]:
                swaps += 1
                target_position = pos[target[i]]
                pos[original[i]] = target_position
                original[target_position] = original[i]
        return swaps

File: java/2471. Minimum Number of Operations to Sort a Binary Tree by Level.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int minimumOperations(TreeNode root) {
        Deque<TreeNode> q = new LinkedList<>();
        q.addLast(root);
        while (!q.isEmpty()) {
            int size = q.size();
            int[] level = new int[size];
            while (size-- > 0) {
                TreeNode current = q.pollFirst();
                level[level.length-size] = current.val;
                level.add(current.val);
                if (current.left != null) q.addFirst(current.left);
                if (current.right != null) q.addFirst(current.right);
            }
            answer += getSwaps(level);
        }
        return answer;
    }
    private int getSwaps(int[] arr) {
        int swaps = 0;
        int[] sorted = arr.clone();
        Arrays.sort(sorted);
        Map<Integer, Integer> positions = new HashMap<>();
        for (int i = 0; i < arr.length; i++) {
            positions.put(arr[i], i);
        }
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] != sorted[i]) {
                swaps++;
                int currentPosition = positions.get(target[i]);
                int currentValue = arr[i];
                arr[i] = sorted[i];
                arr[currentPosition] = currentValue;
                positions.put(currentValue, currentPosition);
            }
        }
        return swaps;
    }
}
